### PR TITLE
texinfo: 7.0.3 -> 7.1

### PR DESCRIPTION
**Pull request by afh**
## Description of changes

Diff: http://git.savannah.gnu.org/cgit/texinfo.git/diff/?id=texinfo-7.0.3&id2=texinfo-7.1

Changelog: http://git.savannah.gnu.org/cgit/texinfo.git/tree/ChangeLog?h=texinfo-7.1

## Things done

Add texinfo 7.1 and make it the default.

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -20289,7 +20289,8 @@ with pkgs;
   texinfo6_5 = callPackage ../development/tools/misc/texinfo/6.5.nix { }; # needed for allegro
   texinfo6_7 = callPackage ../development/tools/misc/texinfo/6.7.nix { }; # needed for gpm, iksemel and fwknop
   texinfo6 = callPackage ../development/tools/misc/texinfo/6.8.nix { };
-  texinfo7 = callPackage ../development/tools/misc/texinfo/7.0.nix { };
+  texinfo7_0 = callPackage ../development/tools/misc/texinfo/7.0.nix { };
+  texinfo7 = callPackage ../development/tools/misc/texinfo/7.1.nix { };
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:20293]: Let's be more explicit:

```suggestion
  texinfo7_0 = callPackage ../development/tools/misc/texinfo/7.0.nix { };
  texinfo7_1 = callPackage ../development/tools/misc/texinfo/7.1.nix { };
  texinfo = texinfo7_1;
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (afh) [reply pkgs/top-level/all-packages.nix:20293]: @AndersonTorres would like to commit your suggestion, yet the message I receive over and over again when trying to do so is: `This diff has recently been updated. Refresh and try again.`

Are you okay with me making the change locally and including you suggestion in a single commit?

Comment (AndersonTorres) [reply pkgs/top-level/all-packages.nix:20293]: OK to me.

Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix]: > Are you okay with me making the change locally and including you suggestion in a single commit?

It should have been squashed anyway

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -20289,8 +20289,9 @@ with pkgs;
   texinfo6_5 = callPackage ../development/tools/misc/texinfo/6.5.nix { }; # needed for allegro
   texinfo6_7 = callPackage ../development/tools/misc/texinfo/6.7.nix { }; # needed for gpm, iksemel and fwknop
   texinfo6 = callPackage ../development/tools/misc/texinfo/6.8.nix { };
-  texinfo7 = callPackage ../development/tools/misc/texinfo/7.0.nix { };
-  texinfo = texinfo7;
+  texinfo7_0 = callPackage ../development/tools/misc/texinfo/7.0.nix { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:20292]: Let me ask a naive question: do we even need 7.0 right now?

**Comment by trofi** [comment]
Not sure if it's just `texonfo-7.1` update or some extra patches I have locally, but I am seeing the following failure against this PR on `ffmpeg-headless`:

```
ffmpeg-headless> makeinfo: error parsing ./doc/t2h.pm: Undefined subroutine &Texinfo::Config::set_from_init_file called at ./doc/t2h.pm line 24.
ffmpeg-headless> POD    doc/libavfilter.pod
ffmpeg-headless> make: *** [doc/Makefile:69: doc/libavformat.html] Error 1 shuffle=3953956671
ffmpeg-headless> make: *** Waiting for unfinished jobs....
```

Does it work for you by chance?

**Comment by afh** [comment]
Thanks for testing, @trofi. I'm seeing the same issue as you are when trying to run `ffmpeg-headless` in combination with the changes in this PR.

It seems the subroutine in question (among others) was prefixed with `texinfo_` (see [this commit](https://git.savannah.gnu.org/cgit/texinfo.git/commit/?id=b7119e08ecda10e0c3a9303893294d169c4bff90)) in September 2021 along with other [refactorings](https://git.savannah.gnu.org/cgit/texinfo.git/commit/?id=a5bbbd3a859fda73980684bc5228bbb542a29691) that break producing the FFmpeg docs.

Applying the patch below in `pkgs/development/libraries/ffmpeg/generic.nix` fixes the issue. What is the best course of action from your point of view?
Report the patch to [FFmpeg upstream](https://github.com/FFmpeg/FFmpeg) and add it to the ffmpeg nix package for the time being?

<details><summary><code>fix-t2h-for-texinfo-7_1.patch</code></summary>

```patch
diff --git a/doc/t2h.pm b/doc/t2h.pm
index d07d974..fc85e04 100644
--- a/doc/t2h.pm
+++ b/doc/t2h.pm
@@ -21,7 +21,7 @@
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 
 # no navigation elements
-set_from_init_file('HEADERS', 0);
+texinfo_set_from_init_file('HEADERS', 0);
 
 sub ffmpeg_heading_command($$$$$)
 {
@@ -55,8 +55,9 @@ sub ffmpeg_heading_command($$$$$)
         $element = $command->{'parent'};
     }
     if ($element) {
-        $result .= &{$self->{'format_element_header'}}($self, $cmdname,
-                                                       $command, $element);
+        $result .= &{$self->formatting_function('format_element_header')}($self,
+                                                       $cmdname, $command,
+                                                       $element);
     }
 
     my $heading_level;
@@ -112,8 +113,8 @@ sub ffmpeg_heading_command($$$$$)
                 $cmdname
                     = $Texinfo::Common::level_to_structuring_command{$cmdname}->[$heading_level];
             }
-            $result .= &{$self->{'format_heading_text'}}(
-                        $self, $cmdname, $heading,
+            $result .= &{$self->formatting_function('format_heading_text')}(
+                        $self, $cmdname, undef, $heading,
                         $heading_level +
                         $self->get_conf('CHAPTER_HEADER_LEVEL') - 1, $command);
         }
@@ -127,22 +128,22 @@ foreach my $command (keys(%Texinfo::Common::sectioning_commands), 'node') {
 }
 
 # determine if texinfo is at least version 6.8
-my $program_version_num = version->declare(get_conf('PACKAGE_VERSION'))->numify;
+my $program_version_num = version->declare(texinfo_get_conf('PACKAGE_VERSION'))->numify;
 my $program_version_6_8 = $program_version_num >= 6.008000;
 
 # print the TOC where @contents is used
 if ($program_version_6_8) {
-    set_from_init_file('CONTENTS_OUTPUT_LOCATION', 'inline');
+    texinfo_set_from_init_file('CONTENTS_OUTPUT_LOCATION', 'inline');
 } else {
-    set_from_init_file('INLINE_CONTENTS', 1);
+    texinfo_set_from_init_file('INLINE_CONTENTS', 1);
 }
 
 # make chapters <h2>
-set_from_init_file('CHAPTER_HEADER_LEVEL', 2);
+texinfo_set_from_init_file('CHAPTER_HEADER_LEVEL', 2);
 
 # Do not add <hr>
-set_from_init_file('DEFAULT_RULE', '');
-set_from_init_file('BIG_RULE', '');
+texinfo_set_from_init_file('DEFAULT_RULE', '');
+texinfo_set_from_init_file('BIG_RULE', '');
 
 # Customized file beginning
 sub ffmpeg_begin_file($$$)
@@ -159,7 +160,7 @@ sub ffmpeg_begin_file($$$)
     my ($title, $description, $encoding, $date, $css_lines,
         $doctype, $bodytext, $copying_comment, $after_body_open,
         $extra_head, $program_and_version, $program_homepage,
-        $program, $generator) = $self->_file_header_informations($command);
+        $program, $generator) = $self->_file_header_information($command);
 
     my $links = $self->_get_links ($filename, $element);
 
@@ -223,7 +224,7 @@ if ($program_version_6_8) {
 sub ffmpeg_end_file($)
 {
     my $self = shift;
-    my $program_string = &{$self->{'format_program_string'}}($self);
+    my $program_string = &{$self->formatting_function('format_program_string')}($self);
     my $program_text = <<EOT;
       <p style="font-size: small;">
         $program_string
@@ -244,7 +245,7 @@ if ($program_version_6_8) {
 
 # Dummy title command
 # Ignore title. Title is handled through ffmpeg_begin_file().
-set_from_init_file('USE_TITLEPAGE_FOR_TITLE', 1);
+texinfo_set_from_init_file('USE_TITLEPAGE_FOR_TITLE', 1);
 sub ffmpeg_title($$$$)
 {
     return '';
@@ -262,7 +263,7 @@ sub ffmpeg_float($$$$$)
     my $args = shift;
     my $content = shift;
 
-    my ($caption, $prepended) = Texinfo::Common::float_name_caption($self,
+    my ($caption, $prepended) = Texinfo::Convert::Converter::float_name_caption($self,
                                                                 $command);
     my $caption_text = '';
     my $prepended_text;
@@ -335,21 +336,21 @@ sub ffmpeg_float($$$$$)
             $caption->{'args'}->[0], 'float caption');
     }
     if ($prepended_text.$caption_text ne '') {
-        $prepended_text = $self->_attribute_class('div','float-caption'). '>'
+        $prepended_text = $self->html_attribute_class('div', ['float-caption']). '>'
                 . $prepended_text;
         $caption_text .= '</div>';
     }
-    my $html_class = '';
+    my @html_class = ();
     if ($prepended_save =~ /NOTE/) {
-        $html_class = 'info';
+        push @html_class, 'info';
         $prepended_text = '';
         $caption_text   = '';
     } elsif ($prepended_save =~ /IMPORTANT/) {
-        $html_class = 'warning';
+        push @html_class, 'warning';
         $prepended_text = '';
         $caption_text   = '';
     }
-    return $self->_attribute_class('div', $html_class). '>' . "\n" .
+    return $self->html_attribute_class('div', \@html_class). '>' . "\n" .
         $prepended_text . $caption_text . $content . '</div>';
 }
```

</details>

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:20292]: Then I would suggest to use it right away and add a comment when we can switch to 7.1

Comment (afh) [reply pkgs/top-level/all-packages.nix:20292]: @SuperSandro2000 What are your thoughts around removing `texinfo7_0` in favor of `texinfo7_1` and using the [patch below](https://github.com/NixOS/nixpkgs/pull/262276#issuecomment-1784209023)?

Comment (AndersonTorres) [reply pkgs/top-level/all-packages.nix:20292]: Do not remove 7.0 if it is still needed.

Or at least remove it in a separate PR.

**Comment by afh** [comment]
ℹ️ Testing the proposed changes here with [ffmpeg 6.1](https://ffmpeg.org/download.html#release_6.1)

**Comment by trofi** [comment]
Sure! I applied both to `staging` and got `ffmped_6` failure as:

```
$ nix build --no-link -f. ffmpeg_6 --keep-going

error: builder for '/nix/store/nl2h321i26ccgmdbfwi9br1qh0m4dgkk-ffmpeg-headless-6.1.drv' failed with exit code 1;
       last 10 log lines:
       > Running phase: configurePhase
       > patching script interpreter paths in ./configure
       > configure flags: --disable-static --prefix=/nix/store/8lf8dsd42v79p9z0mgwnifhyybag32jh-ffmpeg-headless-6.1 --target_os=linux --arch=x86_64 --pkg-config=pkg-config --enable-gpl --enable-version3 --disable-nonfree --enable-shared --enable-pic --disable-small --enable-runtime-cpudetect --disable-gray --enable-swscale-alpha --enable-hardcoded-tables --enable-safe-bitstream-reader --enable-pthreads --disable-w32threads --disable-os2threads --enable-network --enable-pixelutils --datadir=/nix/store/svy6ssg03rz40hz743xwznhsam4hidgg-ffmpeg-headless-6.1-data/share/ffmpeg --enable-ffmpeg --disable-ffplay --enable-ffprobe --bindir=/nix/store/7rhxw2llf53jwcpf7hqr78aal6kw101x-ffmpeg-headless-6.1-bin/bin --enable-avcodec --enable-avdevice --enable-avfilter --enable-avformat --enable-avutil --enable-postproc --enable-swresample --enable-swscale --libdir=/nix/store/c596idk22gd4wni7krvhkx6qfwzbxpsl-ffmpeg-headless-6.1-lib/lib --incdir=/nix/store/gaqkwwq9pmqaa6zljrbn2bxn07zf9sdl-ffmpeg-headless-6.1-dev/include --enable-doc --enable-htmlpages --enable-manpages --mandir=/nix/store/mi6x2ykpvb3yrr7jgc3725v31k74d763-ffmpeg-headless-6.1-man/share/man --enable-podpages --enable-txtpages --docdir=/nix/store/frg6p28dav1i2l7fbx42fn6758b2pcmn-ffmpeg-headless-6.1-doc/share/doc/ffmpeg --enable-alsa --enable-bzlib --disable-libcelt --disable-cuda --disable-cuda-llvm --enable-libdav1d --disable-libfdk-aac --disable-libflite --enable-fontconfig --enable-libfreetype --disable-frei0r --disable-libfribidi --disable-libgme --enable-gnutls --disable-libgsm --disable-ladspa --enable-libmp3lame --disable-libaom --enable-libass --disable-libbluray --disable-libbs2b --disable-libdc1394 --enable-libdrm --enable-iconv --disable-libjack --disable-libmfx --disable-libmodplug --disable-libmysofa --enable-libopus --disable-librsvg --enable-libsrt --enable-libssh --disable-libtensorflow --enable-libtheora --enable-libv4l2 --enable-v4l2-m2m --enable-vaapi --disable-vdpau --enable-libvorbis --disable-libvmaf --enable-libvpx --disable-libwebp --disable-xlib --disable-libxcb --disable-libxcb-shm --disable-libxcb-xfixes --disable-libxcb-shape --disable-libxml2 --enable-lzma --enable-cuvid --enable-nvdec --enable-nvenc --disable-openal --disable-opencl --disable-libopencore-amrnb --disable-opengl --disable-libopenh264 --disable-libopenjpeg --disable-libopenmpt --disable-libpulse --disable-librav1e --enable-libsvtav1 --disable-librtmp --disable-sdl2 --enable-libsoxr --enable-libspeex --disable-libvidstab --disable-libvo-amrwbenc --enable-libx264 --enable-libx265 --disable-libxavs --enable-libxvid --disable-libzmq --enable-libzimg --enable-zlib --disable-vulkan --disable-libglslang --disable-libsmbclient --disable-debug --enable-optimizations --disable-extra-warnings --disable-stripping
       > ERROR: cuvid requested, but not all dependencies are satisfied: ffnvcodec
       >
       > If you think configure made a mistake, make sure you are using the latest
       > version from Git.  If the latest version fails, report the problem to the
       > ffmpeg-user@ffmpeg.org mailing list or IRC #ffmpeg on irc.libera.chat.
       > Include the log file "ffbuild/config.log" produced by configure as this will help
       > solve the problem.
       For full logs, run 'nix log /nix/store/nl2h321i26ccgmdbfwi9br1qh0m4dgkk-ffmpeg-headless-6.1.drv'.
```

Does it build for you?

**Review by AndersonTorres — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
Please remove the merge commit from the PR. It could cause issues if a merge conflict is on it.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix]: We can also do this later.

**Comment by afh** [comment]
Of course. Thanks for calling this out, @SuperSandro2000, the merge commit has now been removed.

**Comment by SuperSandro2000** [comment]
The eval failure is probably because texinfo is so deep into the tree and causes to many rebuilds.

@ofborg eval

**Review by AndersonTorres — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by afh** [comment]
Friendly ping to @AndersonTorres and @SuperSandro2000 on this 🙂

**Comment by AndersonTorres** [comment]
Let's send it to the Discourse thread.

**Comment by kirillrdy** [comment]
@afh please add link to Diff / change log to commit message and PR, eg 44bee06d90d444d26903fb967811f5c0b29c2ab1

**Comment by afh** [comment]
Sure think @kirillrdy, the commit message and PR description have been updated.

Not that the diff link seems to work, but stalls for me after some time, not sure if the diff is too large…

**Comment by kirillrdy** [comment]
> Sure thing, @kirillrdy, the commit message and PR description have been updated.
> 
> Not that the diff link seems to work, but stalls for me after some time, not sure if the diff is too large…

for major releases link to chagelog/release log is sufficient. as you stated, if diff is too large, then its probably not that useful

**Comment by kirillrdy** [comment]
@infinisil sorry to bother you, can you recommend what to do about by-name check in this case ?

**Comment by afh** [comment]
Would moving texinfo into the `pkgs/by-name` tree be something people would find acceptable? I'd be happy to draft a PR, if folks see value in it.

**Comment by AndersonTorres** [comment]
No, not good. Use this instead:

https://github.com/NixOS/nixpkgs/pull/287918

https://github.com/NixOS/nixpkgs/pull/287918#issuecomment-1939243084

**Comment by afh** [comment]
Very interesting, @AndersonTorres, thanks for sharing. Will this also address/fix the by-name check failures?

**Comment by afh** [comment]
Also could you provide more context, please, on why moving texinfo into the by-name tree is bad? I'm genuinely curious.

**Comment by afh** [comment]
@AndersonTorres, I have some local changes for texinfo in the same vein as the PRs you referenced. Should this PR be re-used for that or would a new (draft?) PR be more appropriate?

**Comment by afh** [comment]
Eager to learn more about nixpkgs and motivated by [@AndersonTorres' comment](https://github.com/NixOS/nixpkgs/pull/262276#issuecomment-1950275968) I drafted #289690. Since you, @AndersonTorres and @kirillrdy, have been so kind to share your knowledge and expertise on this PR recently, maybe you'd like to leave a feedback on the draft PR as well; I'd appreciate it greatly.

**Comment by afh** [comment]
Closing as #289690 got merged.

**Pull request closed without merging**